File: frontend/src/services/article-version.service.ts
import { api } from './api'
import type {
  ArticleVersionCreateRequestDto,
  ArticleVersionDto,
  BaseResponseDto,
  VersionSupplementAttachRequestDto,
  VersionSupplementDto,
} from '../models'

export const articleVersionService = {
  createVersion: async (
    articleId: string,
    payload: ArticleVersionCreateRequestDto,
  ): Promise<BaseResponseDto<ArticleVersionDto>> => {
    const response = await api.post<BaseResponseDto<ArticleVersionDto>>(
      `/articles/${articleId}/versions`,
      payload,
    )
    return response.data
  },

  listVersions: async (articleId: string): Promise<BaseResponseDto<ArticleVersionDto[]>> => {
    const response = await api.get<BaseResponseDto<ArticleVersionDto[]>>(
      `/articles/${articleId}/versions`,
    )
    return response.data
  },

  attachSupplement: async (
    articleId: string,
    version: number,
    payload: VersionSupplementAttachRequestDto,
  ): Promise<BaseResponseDto<ArticleVersionDto>> => {
    const response = await api.post<BaseResponseDto<ArticleVersionDto>>(
      `/articles/${articleId}/versions/${version}/attachments`,
      payload,
    )
    return response.data
  },

  listSupplements: async (
    articleId: string,
    version: number,
  ): Promise<BaseResponseDto<VersionSupplementDto[]>> => {
    const response = await api.get<BaseResponseDto<VersionSupplementDto[]>>(
      `/articles/${articleId}/versions/${version}/supplements`,
    )
    return response.data
  },

  mainDownloadUrl: async (
    articleId: string,
    version: number,
  ): Promise<BaseResponseDto<string>> => {
    const response = await api.get<BaseResponseDto<string>>(
      `/articles/${articleId}/versions/${version}/main/download-url`,
    )
    return response.data
  },

  supplementDownloadUrl: async (
    articleId: string,
    version: number,
    attachmentId: string,
  ): Promise<BaseResponseDto<string>> => {
    const response = await api.get<BaseResponseDto<string>>(
      `/articles/${articleId}/versions/${version}/supplements/${attachmentId}/download-url`,
    )
    return response.data
  },

  downloadMainFile: async (articleId: string, version: number): Promise<Blob> => {
    const response = await api.get(`/articles/${articleId}/pdf`, {
      params: { version },
      responseType: 'blob',
    })
    return response.data as Blob
  },

  downloadSupplementFile: async (attachmentId: string): Promise<Blob> => {
    const response = await api.get(`/attachments/${attachmentId}/download`, {
      responseType: 'blob',
    })
    return response.data as Blob
  },
}

File: backend/src/main/kotlin/com/example/researchreview/services/ArticleVersionService.kt
package com.example.researchreview.services

import com.example.researchreview.dtos.ArticleVersionCreateRequestDto
import com.example.researchreview.dtos.ArticleVersionDto
import com.example.researchreview.dtos.VersionSupplementAttachRequestDto
import com.example.researchreview.dtos.VersionSupplementDto

interface ArticleVersionService {
    fun createVersion(articleId: String, request: ArticleVersionCreateRequestDto): ArticleVersionDto
    fun listVersions(articleId: String): List<ArticleVersionDto>
    fun attachSupplement(articleId: String, version: Int, request: VersionSupplementAttachRequestDto): ArticleVersionDto
    fun listSupplements(articleId: String, version: Int): List<VersionSupplementDto>
    fun mainDownloadUrl(articleId: String, version: Int, expirationSeconds: Long = 900): String
    fun supplementDownloadUrl(articleId: String, version: Int, attachmentId: String, expirationSeconds: Long = 900): String
}
